### PR TITLE
Implement BYON CI for OpenStack platform

### PR DESCRIPTION
**Pull request by EmilienM**
This effort is a first iteration over testing different scenarios, all
related to "Bring Your Own Network" feature.

This patch creates 2 new kind of steps (each of them have provision
& deprovision steps):

* One for creating a network in OpenStack, that will be used to deploy
  the OCP cluster. It's called `machinesubnet`. It creates a network,
  a subnet, a router and plug the subnet onto that router so it can
  reach the external networks (Internet).
* One for creating the bastion node, which in fact will act as a proxy.
  This will use Squid and basically help the CI to reach the previously
  created internal network, to deploy and test OCP.
  Note that the Baremetal team is also working on this topic and we will
  probably revisit our steps / scripts to align how we do.

Then, this patch creats a new workflow, called
`openshift-e2e-openstack-byon` and sets the CONFIG_TYPE to `byon`.
We will have more config types in the future, for disconnected installs
for example.

And finally, create a job that will run on both the installer and CAPO.

Note: the jobs don't run the test suite because a lot of test don't
support HTTP proxy and therefore fail. Until we find a sane solution,
for now we're satisfied by just deploying an OCP cluster and if it
finishes successfully then the job will pass.

Co-Authored-By: Adolfo Duarte <adduarte@redhat.com>
Co-Authored-By: Martin André <m.andre@redhat.com>

Implements [OSASINFRA-2022](https://issues.redhat.com/browse/OSASINFRA-2022)

### CODE REVIEW

File: ci-operator/step-registry/openshift/e2e/openstack/fipless/openshift-e2e-openstack-fipless-workflow.yaml
Comment (EmilienM) [inline ci-operator/step-registry/openshift/e2e/openstack/fipless/openshift-e2e-openstack-fipless-workflow.yaml:5]: this step fails to find the CLUSTER_NAME:
```
/bin/bash: line 14: /tmp/secret/CLUSTER_NAME: No such file or directory
```

Comment (mandre) [reply ci-operator/step-registry/openshift/e2e/openstack/fipless/openshift-e2e-openstack-fipless-workflow.yaml:5]: CLUSTER_NAME is written to SHARED_DIR in ci-operator/step-registry/openstack/conf/resourcenames/openstack-conf-resourcenames-commands.sh.
We'll need to run the `openstack-provision-machinesubnet` and `openstack-provision-bastionproxy` steps as part of the `ipi-openstack-pre` chain.

File: ci-operator/step-registry/openstack/provision/bastionproxy/openstack-provision-bastionproxy-commands.sh
Comment (EmilienM) [inline ci-operator/step-registry/openstack/provision/bastionproxy/openstack-provision-bastionproxy-commands.sh:84]: TODO: put the IP in the SHARED_DIR, the export can't work.

File: ci-operator/step-registry/gather/audit-logs/gather-audit-logs-commands.sh
Comment (EmilienM) [inline ci-operator/step-registry/gather/audit-logs/gather-audit-logs-commands.sh:20]: I'll move that to a script, and do like baremetal is doing (see above) do just source it.

**Comment by mandre** [comment]
Hmm, the job failed and we weren't able to gather logs...
/test e2e-openstack-fipless

File: ci-operator/step-registry/ipi/install/install/ipi-install-install-commands.sh
Comment (MaysaMacedo) [inline ci-operator/step-registry/ipi/install/install/ipi-install-install-commands.sh:66]: Should proxy-conf be handled only when cluster-type is openstack (line 80)?

File: ci-operator/step-registry/openstack/conf/generateconfig/openstack-conf-generateconfig-commands.sh
Comment (MaysaMacedo) [inline ci-operator/step-registry/openstack/conf/generateconfig/openstack-conf-generateconfig-commands.sh:24]: not in?

**Review by MaysaMacedo — COMMENTED** [review]
Amazing progress!

File: ci-operator/step-registry/ipi/install/install/ipi-install-install-commands.sh
Comment (mandre) [reply ci-operator/step-registry/ipi/install/install/ipi-install-install-commands.sh:66]: I've wondered that as well and do not completely understand why BM doesn't need to set the proxy with above code in the deploy and test stages.

Comment (EmilienM) [reply ci-operator/step-registry/ipi/install/install/ipi-install-install-commands.sh:66]: @andfasano please give your opinion here ^

Comment (andfasano) [reply ci-operator/step-registry/ipi/install/install/ipi-install-install-commands.sh:66]: Currently [baremetal IPI workflows](https://github.com/openshift/release/tree/master/ci-operator/step-registry/baremetalds/) are running directly the tests from [within the Packet instance](https://github.com/openshift/release/blob/75e93ece163ac3f7a98bc39d1b171c422ec82bb3/ci-operator/step-registry/baremetalds/e2e/test/baremetalds-e2e-test-commands.sh#L101-L105).
We're anyhow also currently working on removing this limitation using the same approach proxy-based (see https://issues.redhat.com/browse/MPINSTALL-24) in order to reuse in future directly the `openshift-e2e-test` step (right now we're using a sibling step pretty close to it).

In order to avoid any impact, I'd suggest to enable the proxy l[ater in the case ](https://github.com/openshift/release/pull/19949/commits/1695a2f861455a127a2fe9c46f5ed4a910899534#diff-1a46129e809990d7d8a81a52a2a48bc8258fae24f171dccb14d3bf00301a3b35L97)only for the cluster profiles requiring it: this will be less disruptive and will allow a smooth migration also for the other profiles

Comment (andfasano) [reply ci-operator/step-registry/ipi/install/install/ipi-install-install-commands.sh:66]: See https://github.com/openshift/release/pull/19987

File: ci-operator/step-registry/openstack/conf/generateconfig/openstack-conf-generateconfig-commands.sh
Comment (EmilienM) [reply ci-operator/step-registry/openstack/conf/generateconfig/openstack-conf-generateconfig-commands.sh:24]: ok I'll change it in the next iteration.

File: ci-operator/step-registry/ipi/install/install/ipi-install-install-commands.sh
Comment (mandre) [reply ci-operator/step-registry/ipi/install/install/ipi-install-install-commands.sh:66]: It's great that you're trying to achieve the same thing. We should join forces, and eventually extract the squid config in its own step that we share among platforms.

I'm still not convinced there is a benefit of sourcing the proxy environment variables in a function like you did in [your PR](https://github.com/openshift/release/pull/19987/files#diff-7b4946fc316154f2f0fe2beac2371cf82bc14d8071298890e8a014417e88159fR39). If the job creates a `${SHARED_DIR}/proxy-conf.sh` file, tha't because we need a proxy to connect to the openshift environment and we should make use of it IMO. Enabling the proxy per-platform would also be inconstant with the pattern we use in other steps (must-gather, etc).

File: ci-operator/step-registry/openstack/conf/generateconfig/openstack-conf-generateconfig-commands.sh
Comment (mandre) [reply ci-operator/step-registry/openstack/conf/generateconfig/openstack-conf-generateconfig-commands.sh:24]: I agree it's a bit confusing to use a switch statement like that, but IIUC the code is functionally correct: it exits with an error unless you use CONFIG_TYPE minimal or fipless.

File: ci-operator/step-registry/ipi/install/install/ipi-install-install-commands.sh
Comment (EmilienM) [reply ci-operator/step-registry/ipi/install/install/ipi-install-install-commands.sh:66]: > (...) eventually extract the squid config in its own step that we share among platforms.

I wonder if we could have a step generating the `squid.conf` and put it in ${SHARED_DIR}, that way each platform can deal with it how they want (e.g. baremetal runs squid in a container on their bastion, openstack can just scp the file into their ephemeral proxy node, etc).
What do you think?

Comment (andfasano) [reply ci-operator/step-registry/ipi/install/install/ipi-install-install-commands.sh:66]: > It's great that you're trying to achieve the same thing. We should join forces, and eventually extract the squid config in its own step that we share among platforms.

Agree, the [proxy step](https://github.com/openshift/release/tree/master/ci-operator/step-registry/baremetalds/devscripts/proxy) could be refactored into the `/openshift` folder. 

> 
> I'm still not convinced there is a benefit of sourcing the proxy environment variables in a function like you did in [your PR](https://github.com/openshift/release/pull/19987/files#diff-7b4946fc316154f2f0fe2beac2371cf82bc14d8071298890e8a014417e88159fR39). If the job creates a `${SHARED_DIR}/proxy-conf.sh` file, tha't because we need a proxy to connect to the openshift environment and we should make use of it IMO. Enabling the proxy per-platform would also be inconstant with the pattern we use in other steps (must-gather, etc).

That sounds reasonable, the per-cluster definition it's just a conservative approach to be sure to not break accidentally other profiles (if the proxy step ends up in a chain, it could not immediately evident), bu rehearshals could help in clarifying it.

File: ci-operator/step-registry/openstack/conf/generateconfig/openstack-conf-generateconfig-commands.sh
Comment (mandre) [reply ci-operator/step-registry/openstack/conf/generateconfig/openstack-conf-generateconfig-commands.sh:24]: I'm torn... I find the following more readable but less maintainable:
```sh
if [[ "$CONFIG_TYPE" != "minimal" || "$CONFIG_TYPE" != "fipless"]]; then
    echo "No valid install config type specified. Please check CONFIG_TYPE"
    exit 1
fi
```
WDYT?

Comment (EmilienM) [reply ci-operator/step-registry/openstack/conf/generateconfig/openstack-conf-generateconfig-commands.sh:24]: I would be in favor of keeping the `case` which is easier to read, in my taste.

**Comment by EmilienM** [comment]
I had an issue with the latest jobs where ipi-install wasn't running, so the OCP cluster wasn't deployed...

File: ci-operator/step-registry/openstack/provision/machinesubnet/openstack-provision-machinesubnet-commands.sh
Comment (pierreprinetti) [inline ci-operator/step-registry/openstack/provision/machinesubnet/openstack-provision-machinesubnet-commands.sh:28]: nit: you can fine tune the command output without the need for grep/awk:

```
SUBNET_ID="$(openstack subnet create "${CLUSTER_NAME}-subnet" \
          --network ${NET_ID} \
          --subnet-range ${SUBNET_RANGE} \
          --dns-nameserver ${DNS_IP} \
          --allocation-pool start=${ALLOCATION_POOL_START},end=${ALLOCATION_POOL_END} \
          --format value --column id)"
```

File: ci-operator/step-registry/openstack/provision/machinesubnet/openstack-provision-machinesubnet-commands.sh
Comment (pierreprinetti) [inline ci-operator/step-registry/openstack/provision/machinesubnet/openstack-provision-machinesubnet-commands.sh:20]: nit: you can fine tune the command output without the need for grep/awk:

```
NET_ID="$(openstack network create --format value --column id "${CLUSTER_NAME}-network")"
```

File: ci-operator/step-registry/openstack/provision/machinesubnet/openstack-provision-machinesubnet-commands.sh
Comment (pierreprinetti) [inline ci-operator/step-registry/openstack/provision/machinesubnet/openstack-provision-machinesubnet-commands.sh:35]: nit: you can fine tune the command output without the need for grep/awk with `-f value -c id` (see above)

**Review by pierreprinetti — COMMENTED** [review]
That's a great piece of work.

Before I look deeper into the actual logic, I wanted to quickly make available to you a couple comments. If in a rush, just consider this one: conditional branches (`if $fipless then skip`) inside steps could be avoided by creating separate provision/deprovision chains specifically for fipless.

We can always find a clever solution for composable chains as a follow-up, if we don't like the repetition that two separate chains would represent.

File: ci-operator/step-registry/openshift/e2e/test/openshift-e2e-test-commands.sh
Comment (EmilienM) [reply ci-operator/step-registry/openshift/e2e/test/openshift-e2e-test-commands.sh:32]: ack, but this was copy/pasted from another snippet. I'll change it in the other places if you want.

File: ci-operator/step-registry/ipi/install/install/ipi-install-install-commands.sh
Comment (EmilienM) [reply ci-operator/step-registry/ipi/install/install/ipi-install-install-commands.sh:66]: I'll close this thread for now, as I think that the `squid.conf` generation is something we can share/optimize in a second phase of this effort.

File: ci-operator/step-registry/openstack/conf/createfips/openstack-conf-createfips-commands.sh
Comment (EmilienM) [reply ci-operator/step-registry/openstack/conf/createfips/openstack-conf-createfips-commands.sh:7]: So if we do that, we have to duplicate `ipi-conf-openstack-chain` for fipless (and maintain it) and then its parents, `ipi-openstack-pre-chain` and `ipi-openstack-post-chain`. I honestly don't think we should go down that path.

Also, there is no way to put conditionals in the steps syntax and I was told it is only done via bash today.

**Comment by EmilienM** [comment]
@pierreprinetti sorry I made a mistake in my previous push, I was working on another PR. I rectified and applied your change again. Thanks a lot for your help on this PR. Let me know if I missed something!

**Comment by EmilienM** [comment]
@pierreprinetti this is failing here: https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_release/19949/rehearse-19949-pull-ci-openshift-cluster-api-provider-openstack-master-e2e-openstack-fipless/1415874451836243968/artifacts/e2e-openstack-fipless/openstack-conf-generateconfig/build-log.txt
This is because we create the API_IP in ci-operator/step-registry/openstack/conf/createfips/openstack-conf-createfips-commands.sh and only if we're not in fipless.

File: ci-operator/step-registry/openstack/conf/createfips/openstack-conf-createfips-commands.sh
Comment (EmilienM) [inline ci-operator/step-registry/openstack/conf/createfips/openstack-conf-createfips-commands.sh:17]: @pierreprinetti the file won't be created in FIPless mode, we need to create it elsewhere as well.

File: ci-operator/step-registry/openstack/provision/bastionproxy/openstack-provision-bastionproxy-commands.sh
Comment (MaysaMacedo) [inline ci-operator/step-registry/openstack/provision/bastionproxy/openstack-provision-bastionproxy-commands.sh:111]: leftovers?

Comment (EmilienM) [reply ci-operator/step-registry/openstack/provision/bastionproxy/openstack-provision-bastionproxy-commands.sh:111]: yes, I'll clean ip

**Comment by pierreprinetti** [comment]
openshift-installer-master-e2e-openstack-fipless passed.
Now I'd like to see an OpenStack non-filpless job pass as well. Pretty please, pj-rehearse?

/test pj-rehearse

**Comment by pierreprinetti** [comment]
openstack-parallel passed.
/lgtm

**Comment by pierreprinetti** [comment]
/assign @crawford @fedosin @honza

like the bot suggests.

**Comment by EmilienM** [comment]
> openstack-parallel passed.

FYI the job that we care in this patch is named `e2e-openstack-fipless` and passed.

**Comment by pierreprinetti** [comment]
@EmilienM https://github.com/openshift/release/pull/19949#issuecomment-882814546

**Review by Fedosin — APPROVED** [review]
/approve
/hold

@EmilienM feel free to remove "hold" when you get baremetal team confirmation

**Comment by EmilienM** [comment]
/hold cancel
the error seen in other jobs should not be related to this patch.

File: ci-operator/jobs/openshift/cluster-api-provider-openstack/openshift-cluster-api-provider-openstack-master-presubmits.yaml
Comment (wking) [inline ci-operator/jobs/openshift/cluster-api-provider-openstack/openshift-cluster-api-provider-openstack-master-presubmits.yaml:69]: you probably want `always_run: false` with a [`skip_if_only_changed` or `run_if_changed` setting][1] to avoid running this on docs bumps and the like.

[1]: https://docs.ci.openshift.org/docs/how-tos/contributing-openshift-release/#tolerated-changes-to-generated-jobs

File: ci-operator/step-registry/openstack/conf/createfips/openstack-conf-createfips-ref.yaml
Comment (wking) [reply ci-operator/step-registry/openstack/conf/createfips/openstack-conf-createfips-ref.yaml:14]: I'm agnostic about removing steps from chains vs. having this env var, but if we keep the env var, can we document the allowed values?  Currently looks like `fipless` and `minimal`?  With some nodes about what those values mean?  For precedent, see [here][1].

[1]: https://github.com/openshift/release/blob/ceaa516be701fd4daaa0fd6bf620af4f88a30032/ci-operator/step-registry/openshift/e2e/test/openshift-e2e-test-ref.yaml#L12-L33

File: ci-operator/jobs/openshift/cluster-api-provider-openstack/openshift-cluster-api-provider-openstack-master-presubmits.yaml
Comment (EmilienM) [reply ci-operator/jobs/openshift/cluster-api-provider-openstack/openshift-cluster-api-provider-openstack-master-presubmits.yaml:69]: oh, great suggestion. Thanks. I'll push over in the next minute.

**Comment by EmilienM** [comment]
I had a suggestion to rename `openshift-e2e-openstack-fipless` to something else because it could be confused with FIPS (instead of FIP, as Floating IP). I'll wait for feedback on that and push for a change. I personally suggest we use `openshift-e2e-openstack-proxy` which would reflect what this job actually does, an install via a proxy.

**Comment by EmilienM** [comment]
@wking Thanks for your comments, I think I addressed them all.

File: ci-operator/step-registry/gather/extra/gather-extra-commands.sh
Comment (vrutkovs) [inline ci-operator/step-registry/gather/extra/gather-extra-commands.sh:34]: Other step changes would require additional approves, should these be moved to a different PR?

File: ci-operator/step-registry/openshift/e2e/openstack/byon/openshift-e2e-openstack-byon-workflow.yaml
Comment (vrutkovs) [inline ci-operator/step-registry/openshift/e2e/openstack/byon/openshift-e2e-openstack-byon-workflow.yaml:10]: I'm pretty sure `openshift-e2e-test` runs well with proxy, at least minimal setup does.

Comment (EmilienM) [reply ci-operator/step-registry/openshift/e2e/openstack/byon/openshift-e2e-openstack-byon-workflow.yaml:10]: would you mind if I try as a follow up patch? I'll test it right after this patch has landed. Thanks!

File: ci-operator/jobs/openshift/cluster-api-provider-openstack/openshift-cluster-api-provider-openstack-master-presubmits.yaml
Comment (EmilienM) [reply ci-operator/jobs/openshift/cluster-api-provider-openstack/openshift-cluster-api-provider-openstack-master-presubmits.yaml:69]: done! thanks for the suggestion.

File: ci-operator/step-registry/gather/extra/gather-extra-commands.sh
Comment (EmilienM) [reply ci-operator/step-registry/gather/extra/gather-extra-commands.sh:34]: I'm fine moving it to another PR, but this is a minor change, that brings consistency in shellcheck. If I struggle to get reviews then I'll remove the change in that file.

File: ci-operator/step-registry/openshift/e2e/openstack/byon/openshift-e2e-openstack-byon-workflow.yaml
Comment (vrutkovs) [reply ci-operator/step-registry/openshift/e2e/openstack/byon/openshift-e2e-openstack-byon-workflow.yaml:10]: Sure, that sounds good

**Comment by e-tienne** [comment]
> I had a suggestion to rename `openshift-e2e-openstack-fipless` to something else because it could be confused with FIPS (instead of FIP, as Floating IP). I'll wait for feedback on that and push for a change. I personally suggest we use `openshift-e2e-openstack-proxy` which would reflect what this job actually does, an install via a proxy.

I did stumble on some code / docs using `fip` or `FIP` and it was also confusing for me since we already use `FIPS` elsewhere in the code for the security standard. I think it's a good idea to change the name, or at minimum use Floating IP instead to clarify the term for the unitiated.

My understanding is that we typically use the `proxy` term in this repository when we use and outgoing [proxy like in the OpenShift product docs ](https://docs.openshift.com/container-platform/4.7/installing/installing_aws/installing-restricted-networks-aws-installer-provisioned.html#installation-configure-proxy_installing-restricted-networks-aws-installer-provisioned)

**Comment by EmilienM** [comment]
> I did stumble on some code / docs using `fip` or `FIP` and it was also confusing for me since we already use `FIPS` elsewhere in the code for the security standard. I think it's a good idea to change the name, or at minimum use Floating IP instead to clarify the term for the unitiated.

Right, that's why we renamed it.

> My understanding is that we typically use the `proxy` term in this repository when we use and outgoing [proxy like in the OpenShift product docs ](https://docs.openshift.com/container-platform/4.7/installing/installing_aws/installing-restricted-networks-aws-installer-provisioned.html#installation-configure-proxy_installing-restricted-networks-aws-installer-provisioned)

We'll have a `proxy` job later, where it used httpProxy, etc to deploy OCP.

File: ci-operator/jobs/openshift/cluster-api-provider-openstack/openshift-cluster-api-provider-openstack-master-presubmits.yaml
Comment (wking) [reply ci-operator/jobs/openshift/cluster-api-provider-openstack/openshift-cluster-api-provider-openstack-master-presubmits.yaml:69]: this doesn't have to get pulled in with `/test ...`, because it has a `run_if_changed` block.  So whether it is `optional: true` or not is an orthogonal decision based on whether you want it to block when it does run.  E.g. the CVO has a number of jobs that run whenever non-doc code changes which we want to be blockers.

File: ci-operator/jobs/openshift/cluster-api-provider-openstack/openshift-cluster-api-provider-openstack-master-presubmits.yaml
Comment (wking) [inline ci-operator/jobs/openshift/cluster-api-provider-openstack/openshift-cluster-api-provider-openstack-master-presubmits.yaml:82]: No need for inverse regexps anymore; see `skip_if_only_changed`, e.g. #19409.

File: ci-operator/step-registry/openstack/conf/creatednsrecords/openstack-conf-creatednsrecords-ref.yaml
Comment (wking) [inline ci-operator/step-registry/openstack/conf/creatednsrecords/openstack-conf-creatednsrecords-ref.yaml:15]: nit: `wouln't` -> `won't` or `will not` or whatever.

File: ci-operator/step-registry/openstack/provision/machinesubnet/openstack-provision-machinesubnet-ref.yaml
Comment (wking) [inline ci-operator/step-registry/openstack/provision/machinesubnet/openstack-provision-machinesubnet-ref.yaml:47]: nit: These `${SHARED_DIR}` files are effectively an API for steps to communicate.  If it's always going to be the same owner pool for all steps involved with the API, it's not really a problem to leave the API undocumented.  But if you might someday have different maintainers for different steps using the API, or if you think you might have forgotton what the API was by the next time you come back to work on these files, it's good to document the APIs in more detail, like what content is read from what files as step input, and what content goes into what files as step output.

File: ci-operator/jobs/openshift/cluster-api-provider-openstack/openshift-cluster-api-provider-openstack-master-presubmits.yaml
Comment (EmilienM) [reply ci-operator/jobs/openshift/cluster-api-provider-openstack/openshift-cluster-api-provider-openstack-master-presubmits.yaml:82]: ok

File: ci-operator/step-registry/openstack/conf/creatednsrecords/openstack-conf-creatednsrecords-ref.yaml
Comment (EmilienM) [reply ci-operator/step-registry/openstack/conf/creatednsrecords/openstack-conf-creatednsrecords-ref.yaml:15]: ok I'll address it in a future PR if possible.

File: ci-operator/step-registry/openstack/provision/machinesubnet/openstack-provision-machinesubnet-ref.yaml
Comment (EmilienM) [reply ci-operator/step-registry/openstack/provision/machinesubnet/openstack-provision-machinesubnet-ref.yaml:47]: ack, I'll address it in a future PR if possible.

**Comment by EmilienM** [comment]
/hold cancel
we agreed on slack with @wking  that I'll address nits on a follow-up.
Thanks everyone!

File: ci-operator/step-registry/openshift/e2e/openstack/byon/openshift-e2e-openstack-byon-workflow.yaml
Comment (EmilienM) [reply ci-operator/step-registry/openshift/e2e/openstack/byon/openshift-e2e-openstack-byon-workflow.yaml:10]: @vrutkovs https://github.com/openshift/release/pull/20577

**Pull request merged**
